ops/model.py

```diff
@@ -1483,7 +1509,26 @@ def set_content(self, content: Dict[str, str]):
         self._validate_content(content)
         if self._id is None:
             self._id = self.get_info().id
-        self._backend.secret_set(typing.cast(str, self.id), content=content)
+
+        # If `_backend.secret_set` has already been called, this call will
+        # overwrite anything done in the previous call (even if it was in a
+        # different event handler, as long as it was in the same Juju hook
+        # context). We want to provide more predictable behavior, so we cache
+        # the values and provide them in subsequent calls.
+        cached_data = self._secret_set_cache[self._id]
+        expire = _calculate_expiry(cached_data['expire']) if 'expire' in cached_data else None
+        # Don't store the actual secret content in memory, but put a sentinel
+        # there to indicate that the content has been set during this hook.
+        cached_data['content'] = object()
```

Comment:
I think GitHub ate my home^H^H^H^Hprevious comment

Is this safe in this scenario:

```py
.set_content(...)  # passed via temp file to juju agent
# temp dir ctx manager exits, removes the temp files

.set_info(...)  # another hook tool invocation
```

Does juju agent combine the value with the info in this case?

If it doesn't I guess we've got no choice to cache the value somewhere, either in memory or in longer-lived temporary directory/file.

Reply:
> Is this safe in this scenario:
> 
> ```python
> .set_content(...)  # passed via temp file to juju agent
> # temp dir ctx manager exits, removes the temp files
> 
> .set_info(...)  # another hook tool invocation
> ```
> 
> Does juju agent combine the value with the info in this case?

No (it never combines), but it reads from the file(s) used in the `set_content` during the `secret-set` call. There's no need for those files to stick around after the `secret-set` is finished. Otherwise our use of temp files wouldn't work - we'd have to clean them up after the hook completed somehow.

Ops does (in this branch) combine the two, though. It gets the content from `secret-get` and passes it back (in brand new temporary files).

It is inefficient to do `set_info()` and then `set_content()` compared with a single `secret-set` call - and you're better off doing the `set_info` first if you can, because then at least you don't have ops writing out to the files more than once (and Juju reading from them). I played around with having a new method that let you do both or extending `set_info` to let you pass content as well, but a new method felt too heavy unless we were going to deprecate the other two, and `set_info` seems confusingly named if it can do content as well. I think we'll want to redesign this eventually, but not any time soon.